### ISSUE TITLE
Use Bundesnetzagentur SMARD API instead of EPEX Spot Web Scraper

### ISSUE THREAD

**Issue by TobiasBraeutigam** [mampfes/ha_epex_spot#29]
Hi, 
thank you for your great component. 

I just want to suggest, that you do not need the EPEX Spot Web Scraper anymore. Instead you could use the SMARD Plattform published by the Bundesnetzagentur. There you get all the data from official sources for free and via API.

SMARD is the information platform of the Federal Network Agency about the German electricity market. All users can access visually and tabularly prepared data at [www.smard.de](http://www.smard.de/). The API documented here provides access to the electricity market data of the Federal Network Agency.

https://www.smard.de/home/marktdaten?marketDataAttributes=%7B%22resolution%22:%22hour%22,%22from%22:1684792800000,%22to%22:1685051999999,%22moduleIds%22:%5B8004169%5D,%22selectedCategory%22:8,%22activeChart%22:true,%22style%22:%22color%22,%22categoriesModuleOrder%22:%7B%7D,%22region%22:%22DE%22%7D

https://github.com/bundesAPI/smard-api

https://smard.api.bund.dev/

I hope it helps.

Best regards,
Tobias

**Comment by mampfes** [comment]
Thank you, this is really intersting information. I wasn't aware of that service. I will try to integrate it.

**Comment by TobiasBraeutigam** [comment]
Hi, here is an working example for loading the data from smard:

```python
import datetime
import requests
import pandas as pd

# Mögliche Filter:
# Stromerzeugung
# 1223 - Braunkohle
# 1224 - Kernenergie
# 1225 - Wind Offshore
# 1226 - Wasserkraft
# 1227 - Sonstige Konventionelle
# 1228 - Sonstige Erneuerbare
# 4066 - Biomasse
# 4067 - Wind Onshore
# 4068 - Photovoltaik
# 4069 - Steinkohle
# 4070 - Pumpspeicher
# 4071 - Erdgas

# Stromverbrauch
# 410 - Gesamt (Netzlast)
# 4359 - Residuallast
# 4387 - Pumpspeicher

# Marktpreis
# 4169 - Deutschland/Luxemburg
# 5078 - Anrainer DE/LU
# 4996 - Belgien
# 4997 - Norwegen 2
# 4170 - Österreich
# 252 - Dänemark 1
# 253 - Dänemark 2
# 254 - Frankreich
# 255 - Italien (Nord)
# 256 - Niederlande
# 257 - Polen
# 258 - Polen
# 259 - Schweiz
# 260 - Slowenien
# 261 - Tschechien
# 262 - Ungarn

# Prognostizierte Erzeugung
# 3791 - Offshore
# 123 - Onshore
# 125 - Photovoltaik
# 715 - Sonstige
# 5097 - Wind und Photovoltaik
# 122 - Gesamt

smard_filter = 4169

# Land / Regelzone / Marktgebiet:
# DE - Land: Deutschland
# AT - Land: Österreich
# LU - Land: Luxemburg
# DE-LU - Marktgebiet: DE/LU (ab 01.10.2018)
# DE-AT-LU - Marktgebiet: DE/AT/LU (bis 30.09.2018)
# 50Hertz - Regelzone (DE): 50Hertz
# Amprion- Regelzone (DE): Amprion
# TenneT - Regelzone (DE): TenneT
# TransnetBW - Regelzone (DE): TransnetBW
# APG - Regelzone (AT): APG
# Creos - Regelzone (LU): Creos
smard_region = "DE-LU"

# Auflösung der Daten:
# hour - Stündlich
# quarterhour - Viertelstündlich
# day - Täglich
# week - Wöchentlich
# month - Monatlich
# year - Jährlich
smard_resolution = "hour"

# Verfügbare Timestamps für Filter, Region und Auflösung
url = f"https://www.smard.de/app/chart_data/{smard_filter}/{smard_region}/index_{smard_resolution}.json"
response = requests.get(url)

# Prüfen, ob der Antrag erfolgreich war
if response.status_code == 200:
    data_dict = response.json()  # Konvertieren Sie die Antwort in ein Python-Diktionsobjekt
else:
    print(f"Anfrage ist fehlgeschlagen mit dem Statuscode {response.status_code}")

smard_start_ts = data_dict['timestamps'][-1]



# Zeitreihendaten nach Filter, Region und Auflösung ab Timestamp
url = f"https://www.smard.de/app/chart_data/{smard_filter}/{smard_region}/{smard_filter}_{smard_region}_{smard_resolution}_{smard_start_ts}.json"
response = requests.get(url)

# Prüfen, ob der Antrag erfolgreich war
if response.status_code == 200:
    data_dict = response.json()  # Konvertieren Sie die Antwort in ein Python-Diktionsobjekt
else:
    print(f"Anfrage ist fehlgeschlagen mit dem Statuscode {response.status_code}")



data_df = pd.DataFrame(data_dict["series"], columns=["date", "value"])
data_df.date = pd.to_datetime(data_df.date, unit='ms')
data_df
```

Result:
![image](https://github.com/mampfes/ha_epex_spot/assets/89102799/6047471b-4fcb-41d5-9f6a-10fde36b9b0a)

**Comment by mampfes** [comment]
Thanks, this is really helpful!

**Comment by mampfes** [comment]
Hm, I am struggling with the returned data...

For the following attributes
```
        smard_filter = 4169
        smard_region = "DE-LU"
        smard_resolution = "hour"
```
I only get market prices for today. The price values for tomorrow are all `null` in the json response. Do you see the same values?

**Comment by TobiasBraeutigam** [comment]
Yes, I have the same issue. I mailed SMARD. The days before ist was still working. Other sources like ENTSO-E do provide the missing data.

Gesendet von Outlook für iOS<https://aka.ms/o0ukef>
________________________________
Von: Steffen Zimmermann ***@***.***>
Gesendet: Thursday, June 8, 2023 8:27:32 PM
An: mampfes/ha_epex_spot ***@***.***>
Cc: TobiasBraeutigam ***@***.***>; Author ***@***.***>
Betreff: Re: [mampfes/ha_epex_spot] Use Bundesnetzagentur SMARD API instead of EPEX Spot Web Scraper (Issue #29)


Hm, I am struggling with the returned data...

For the following attributes

        smard_filter = 4169
        smard_region = "DE-LU"
        smard_resolution = "hour"


I only get market prices for today. The price values for tomorrow are all null in the json response. Do you see the same values?

**Comment by TobiasBraeutigam** [comment]
The Data is available again:

![image](https://github.com/mampfes/ha_epex_spot/assets/89102799/d311a9f9-37f8-4880-a8b5-ca5de0f43720)

**Comment by ProfDrYoMan** [comment]
Would be great to see this official source being integrated.

**Comment by mampfes** [comment]
fixed by #32
